proposals/ssss-distributed-actor-runtime.md

```diff
@@ -812,99 +981,128 @@ try await greeter.greet(name: "Alice")
+    // for the generic parameters A and B:
+    //
+    // << invocation.recordGenericSubstitution(<runtime type of a>)
+    // << invocation.recordGenericSubstitution(<runtime type of b>)
+    
+    // [1.2] for each argument, synthesize a specialized recordArgument call:
+    try invocation.recordArgument(name)
+      
+    // [1.3] if the target was throwing, record Error.self, 
+    // otherwise do not invoke recordErrorType at all.
+    //
+    // << try invocation.recordErrorType(Error.self)
+      
+    // [1.4] we also record the return type; it may or may not be necessary to
+    // transmit over the wire but if necessary, the system may choose to do so.
+    // 
+    // This call is not made when the return type is Void.
+    try invocation.recordReturnType(String.self)
+      
+    // [1.5] done recording arguments
+    try invocation.doneRecording()
+      
+    // [2] invoke the `remoteCall` method of the actor system
+    return try await self.actorSystem.remoteCall(
+      on: self,
+      target: RemoteCallTarget(...),
+      invocation: invocation,
+      throwing: Never.self, // the target func was not throwing
+      returning: String.self
+    )
   }
 }
 ```
 
-The synthesized thunk is always throwing and asynchronous, this is correct because it is only invoked in situations where we might end up calling the `actorSystem.remoteCall(...)` method, which by necessity is asynchronous and throwing as well. 
+The synthesized thunk is always throwing and asynchronous, this is correct because it is only invoked in situations where we might end up calling the `actorSystem.remoteCall(...)` method, which by necessity is asynchronous and throwing.
+
+The thunk is `nonisolated` because it is a method that can actually run on a *remote* instance, and as such is not allowed to touch any other state than other nonisolated stored properties. This is specifically designed such that the thunk (and actor system are able to access the `id` of the actor (and the `actorSystem` property itself) which is necessary to perform the actual remote message send.
+
+The `nonisolated` aspect of the method has another important role to play: if this invocation happens to be  on a local distributed actor, we do not want to "hop" executors twice, but only once we have confirmed the actor is local and hop to it using the same semantics as we would when performing a normal actor method call. If the instance was remote, we don't need to suspend early at all, and we leave it to the `ActorSystem` to decide when exactly the task will suspend. For example, the system may only suspend the call after it has sent the bytes synchronously over some IPC channel etc. Those semantics when to suspend are highly dependent on the specific underlying transport, and thanks to this approach we allow system implementations to do the right thing, whatever that might be: they can suspend early, late, or even not at all if the call is known to be impossible to succeed.
```

Comment:
reworded: If this invocation were on a local actor, only accessing `nonisolated` state, or for other reasons the hop could be optimized away, we want to keep this ability for the optimizer to do as good of a job as it would for local only actors.

Reply:
fixed the otehrs too